packages/webview_flutter/test/webview_flutter_test.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:webview_flutter/webview_flutter.dart';
+
+void main() {
+  setUpAll(() {
+    SystemChannels.platform_views
+        .setMockMethodCallHandler(_fakePlatformViewsMethodHandler);
+  });
+  testWidgets('Create WebView', (WidgetTester tester) async {
+    await tester.pumpWidget(const WebView());
+  });
+
+  testWidgets('Load url', (WidgetTester tester) async {
+    final int currentViewId = platformViewsRegistry.getNextPlatformViewId();
+    WebViewController controller;
+    await tester.pumpWidget(
+      WebView(
+        onWebViewCreated: (WebViewController webViewController) {
+          controller = webViewController;
+        },
+      ),
+    );
+
+    expect(controller, isNotNull);
+
+    final FakePlatformWebView platformWebView =
+        new FakePlatformWebView(currentViewId + 1);
+
+    controller.loadUrl('https://flutter.io');
+
+    expect(platformWebView.lastUrlLoaded, 'https://flutter.io');
+  });
+
+  testWidgets('Invald urls', (WidgetTester tester) async {
+    final int currentViewId = platformViewsRegistry.getNextPlatformViewId();
+    WebViewController controller;
+    await tester.pumpWidget(
+      WebView(
+        onWebViewCreated: (WebViewController webViewController) {
+          controller = webViewController;
+        },
+      ),
+    );
+
+    expect(controller, isNotNull);
+
+    final FakePlatformWebView platformWebView =
+        new FakePlatformWebView(currentViewId + 1);
+
+    expect(() => controller.loadUrl(null), throwsA(anything));
+    expect(platformWebView.lastUrlLoaded, isNull);
+
+    expect(() => controller.loadUrl(''), throwsA(anything));
+    expect(platformWebView.lastUrlLoaded, isNull);
+
+    // Missing schema.
+    expect(() => controller.loadUrl('flutter.io'), throwsA(anything));
+    expect(platformWebView.lastUrlLoaded, isNull);
+  });
+}
+
+class FakePlatformWebView {
```

Comment:
This way of testing works too, but you could avoid making a Mock class by doing something similar in [firebase_performance](https://github.com/flutter/plugins/blob/2dc4852c5d357bfb76b5c6f5c926c2937923ac03/packages/firebase_performance/test/firebase_performance_test.dart#L21).

We started testing plugins by making the `MethodChannel` public and using the `@visibleForTesting` annotation. It lets us test the actual class instead of having to always make a Fake one for testing.

Reply:
This fake class is faking the Java implementation, the test is testing the real Dart WebView implementation.

Reply:
I was thinking of something along the lines of:

```
    final List<MethodCall> log = <MethodCall>[];
    WebViewController controller;
    
    Future<dynamic> methodCallHandler(MethodCall methodCall) async {
      log.add(methodCall);

      switch (methodCall.method) {
        case 'loadUrl':
          return null;
        default:
          return null;
      }
    }

    testWidgets('loadUrl', (WidgetTester tester) async {
      await tester.pumpWidget(
        WebView(
          onWebViewCreated: (WebViewController webViewController) {
            controller = webViewController;
          },
        ),
      );
      
      expect(controller, isNotNull);

      controller.channel.setMethodCallHandler(methodCallHandler);
      log.clear();

      controller.loadUrl('https://flutter.io');

      expect(log, <Matcher>[
        isMethodCall(
          'loadUrl',
          arguments: 'https://flutter.io',
        )
      ]);
    });
```

Reply:
Following offline discussion we decided we're happy with the fake approach here.